Refactor `EditSecretModal` to use shadcn components

Saeris opened the pull request:
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES

## What kind of change does this PR introduce?

This primarily refactors `EditSecretModal` for the vault secrets page such that it uses shadcn components instead of older deprecated ones. It follows the same implementation pattern as #36972

Additionally it adds some basic UI tests for this component as well as some fixes and improvements to `studio`'s testing setup and utilities.

Noteworthy is that this introduces an extra layer of type safety to `addAPIMock()` inspired by the pattern used with `createClient()` from `openapi-fetch`. It uses the same generated type definitions to provide type information to the response type for each endpoint you may want to mock. Not only will this provide autocomplete and type info for the response objects, it can also potentially highlight changes to the API which would affect existing tests.

This also adds `jsdom-testing-mocks` as a new development dependency to polyfill web animation APIs necessary for the shadcn Dropdown components to be rendered in tests.

Because this is branched off of work on #36972, I'll be leaving this as a draft until that PR is reviewed and merged.

## Additional context

Part of general deprecated component refactor work at the request of @saltcod

Saeris committed:
chore: Refactor `NewAccessTokenButton` to use shadcn components

This brings `NewAccessTokenButton` into conformance with the latest UI patterns of using shadcn based components, zod schemas, and react-hook-form for form submission.

I based this refactor on the implementation of `CreateSecretAPIKeyDialog`, which already made use of the above.

Saeris committed:
chore: Refactor `NewAccessTokenButton` to use shadcn components

This brings `NewAccessTokenButton` into conformance with the latest UI patterns of using shadcn based components, zod schemas, and react-hook-form for form submission.

I based this refactor on the implementation of `CreateSecretAPIKeyDialog`, which already made use of the above.

Saeris committed:
add success and failure toasts

Saeris committed:
add success and failure toasts

Saeris committed:
add smoke test for `NewAccessTokenButton`, fix vitest config

This adds a minimal test for `NewAccessTokenButton` but more importantly fixes a critical bug in the vitest configuration for `studio`.

Because `restoreMocks: true` was set in the config, this meant that the mock implementation for `window.matchMedia` defined in the `tests/vitestSetup.ts` file was being **reset** to an empty function before each test suite was run. While this didn't appear to be a problem before, that appears to be because none of the existing tests had `motion.div` in their component trees. Since `motion.div` calls `addListener` on a media query as part of it's lifecycle, that meant that any test which included it would fail as `addListener` would be undefined in this situation. Removing `restoreMocks: true` from the config results in the desired mocking behavior.

This change is necessary in order to test any component that has the `FormItemLayout` component, as it uses `framer-motion` to animate in error messages for input elements.

Saeris committed:
add smoke test for `NewAccessTokenButton`, fix vitest config

This adds a minimal test for `NewAccessTokenButton` but more importantly fixes a critical bug in the vitest configuration for `studio`.

Because `restoreMocks: true` was set in the config, this meant that the mock implementation for `window.matchMedia` defined in the `tests/vitestSetup.ts` file was being **reset** to an empty function before each test suite was run. While this didn't appear to be a problem before, that appears to be because none of the existing tests had `motion.div` in their component trees. Since `motion.div` calls `addListener` on a media query as part of it's lifecycle, that meant that any test which included it would fail as `addListener` would be undefined in this situation. Removing `restoreMocks: true` from the config results in the desired mocking behavior.

This change is necessary in order to test any component that has the `FormItemLayout` component, as it uses `framer-motion` to animate in error messages for input elements.

Saeris committed:
add warning comment in case of future config regression

Saeris committed:
update test cases, reset form on dialog close, testing setup fixes

This fixes an issue with the polyfills for the testing environment, where a call to `useMutation` would fail as a result of `TransformStream` being set to `null`.

Basic tests for access token creation and form resetting added. Adds `@faker-js/faker` to the studio app devDependencies to generate mock data for `msw` endpoint response.

Adds `shx` to the docs app devDependencies to ensure that the `codegen:examples` script runs cross-platform.

Saeris committed:
update test cases, reset form on dialog close, testing setup fixes

This fixes an issue with the polyfills for the testing environment, where a call to `useMutation` would fail as a result of `TransformStream` being set to `null`.

Basic tests for access token creation and form resetting added. Adds `@faker-js/faker` to the studio app devDependencies to generate mock data for `msw` endpoint response.

Adds `shx` to the docs app devDependencies to ensure that the `codegen:examples` script runs cross-platform.

Saeris committed:
Merge branch 'master' into chore/refactor-new-access-token-button

Saeris committed:
ensure mocked date is a string

Saeris committed:
ensure mocked date is a string

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
update testing setup, pollyfills and add rich type support to addAPIMock

Saeris committed:
update testing setup, pollyfills and add rich type support to addAPIMock

Saeris committed:
add missing changes

Saeris committed:
add missing changes

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Move `EditSecretModal` to be child of `SecretRow`

In the accompanying PR for `DeleteSecretModal`, that one will also be moved over to `SecretRow`

Saeris committed:
Move `EditSecretModal` to be child of `SecretRow`

In the accompanying PR for `DeleteSecretModal`, that one will also be moved over to `SecretRow`

Saeris committed:
Update studio testing setup files

Improves API mocking type safety and polyfills browser APIs necessary to run tests with framer-motion components

Saeris committed:
Update studio testing setup files

Improves API mocking type safety and polyfills browser APIs necessary to run tests with framer-motion components

Saeris committed:
Merge remote-tracking branch 'Saeris/chore/refactor-edit-vault-secret-modal' into chore/refactor-edit-vault-secret-modal

Saeris committed:
fix imports

Saeris committed:
Update test, fix rebase errors

Saeris committed:
remove unnecessary code

Saeris committed:
resolve type errors

Saeris committed:
fix missing listen call for msw, resolve test type error

Saeris committed:
fix missing listen call for msw, resolve test type error

Saeris committed:
fix imports

Saeris committed:
fix imports

Saeris committed:
Merge remote-tracking branch 'Saeris/chore/update-studio-testing-setup' into chore/update-studio-testing-setup

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
remove duplicate test

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/SecretRow.tsx:
```diff
@@ -128,15 +130,16 @@ const SecretRow = ({ secret, onSelectEdit, onSelectRemove }: SecretRowProps) =>
-        </DropdownMenu>
+              </DropdownMenuContent>
+            </DropdownMenu>
+          </TooltipTrigger>
+          {!canManageSecrets && (
+            <TooltipContent side="bottom">
+              You need additional permissions to manage secrets
+            </TooltipContent>
+          )}
+        </Tooltip>
```
rather than having a repetitive message for each menu item, we can just disable the whole menu since all the items require permission to manage the secret

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/SecretsManagement.tsx:
```diff
@@ -81,26 +88,21 @@ export const SecretsManagement = () => {
+                <SelectItem_Shadcn_ value="updated_at" className="text-xs">
                   Updated at
-                </Listbox.Option>
-                <Listbox.Option
-                  id="name"
-                  className="max-w-[180px]"
-                  value="name"
-                  label="Sort by name"
-                >
+                </SelectItem_Shadcn_>
+                <SelectItem_Shadcn_ value="name" className="text-xs">
                   Name
-                </Listbox.Option>
-              </Listbox>
-            </div>
+                </SelectItem_Shadcn_>
+              </SelectContent_Shadcn_>
+            </Select_Shadcn_>
```
I'm not sure it matters but I do want to point out that the select dropdown icon is different between these two implementations

Saeris replied:
additionally, I noticed that when TanStack dev tools are open, opening this select causes a layout shift, so if you ever see that and wonder why, it's not something you need to worry about debugging

saltcod replied:
the selected check icon does look big. I think this needs `portal=false` to fix the layout issue?

![screenshot-2025-07-28-at-17 13 11](https://github.com/user-attachments/assets/da6cfb14-3544-4130-94b3-19328c83a0d6)

Saeris replied:
hmm, that appears to be off. this is what I am seeing on me end:

<img width="867" height="562" alt="image" src="https://github.com/user-attachments/assets/75616458-8ecf-4dc9-b443-91db1371fbc3" />

that layout issue would appear to be consistent with the old vault sort select

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/SecretsManagement.tsx:
```diff
@@ -167,10 +165,6 @@ export const SecretsManagement = () => {
         </div>
       </div>
 
-      <EditSecretModal
-        selectedSecret={selectedSecretToEdit}
-        onClose={() => setSelectedSecretToEdit(undefined)}
-      />
       <DeleteSecretModal
         selectedSecret={selectedSecretToRemove}
         onClose={() => setSelectedSecretToRemove(undefined)}
```
both of these get moved into `SecretRow` because it doesn't really make sense to have this awkward sibling relationship where the dialog trigger is nested, causing state management in the parent. I think it's the kind of wonky coupling of components that we'd want to avoid

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/EditSecretModal.tsx:
```diff
@@ -1,158 +1,214 @@
+      description: secret?.description ?? '',
+      secret: secret?.decryptedSecret ?? data ?? '',
+    },
+    values: {
+      name: secret?.name ?? '',
+      description: secret?.description ?? '',
+      secret: secret?.decryptedSecret ?? data ?? '',
+    },
```
this might seem like unnecessary duplication but I found that this was what was needed to fill in the values that are loaded in asynchronously from `useVaultSecretDecryptedValueQuery`. it's better than the odd hook inside of render function that was here previously

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/__tests__/EditSecretModal.test.tsx:
```diff
@@ -0,0 +1,130 @@
+    await userEvent.clear(descriptionInput)
+    await userEvent.type(valueInput, `qux`)
+
+    fireEvent.click(submitButton)
```
something to note here, which I went into more detail on Slack, is that I had a heck of a time getting form submission to play nice and for this particular component `fireEvent` was the only method I could get to work *in test*

the form works just as you would expect in the browser, but for whatever reason in a testing environment it doesn't like either `userEvent.click` or an `{enter}` on a text input. I wouldn't waste more time trying to debug why this is the case, but make a mental note of this for writing future tests as it may crop up as a problem you'll encounter

and importantly since `fireEvent` isn't something you can `await` on, you will need to have something like the following line to wait for a change in the UI state when the event in question triggers something asynchronous like a fetch call

Saeris commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/__tests__/EditSecretModal.test.tsx:
```diff
@@ -0,0 +1,130 @@
+    }
+  }
+  return (
+    <ProjectContextProvider projectRef="default">
+      <DropdownMenu>
+        <DropdownMenuTrigger asChild>
+          <Button title="Manage Secret" type="text" className="px-1" icon={<MoreVertical />} />
+        </DropdownMenuTrigger>
+        <DropdownMenuContent side="bottom" align="end" className="w-32">
+          <DropdownMenuItem className="space-x-2" onClick={() => setModal(`edit`)}>
+            <Edit3 size="14" />
+            <p>Edit</p>
+          </DropdownMenuItem>
+        </DropdownMenuContent>
+      </DropdownMenu>
+
+      {renderModal()}
+    </ProjectContextProvider>
```
this much boilerplate may not be necessary, but I wanted to ensure that we're getting the right behavior integrating with `DropdownMenu`, where the two popovers can have some complicated mounting/unmounting behaviors if not done correctly

saltcod commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/SecretRow.tsx:
```diff
@@ -46,6 +46,15 @@ const SecretRow = ({ secret, onSelectEdit, onSelectRemove }: SecretRowProps) =>
+        return null
+    }
+  }
 
```
this is valid, but not common I think for this codebase. Could we use inline conditionals or extract to a new component completely?

Saeris replied:
yeah I mean I suppose we could inline this in the JSX with conditionals instead. I'm using the switch statement pattern here based on a solution I came across researching how to combine a radix dialog with a dropdown menu. personally I like that it keeps this whole conditional UI in one place, close handler and all

Saeris replied:
alright then, how about I implement that change in #37386 then? just to avoid needing to repeat the change across multiple PRs, since we're needing to review and merge them in order anyways

saltcod reviewed (commented):
Left a couple of nits, but tested on staging and adding/changing/deleting secrets works as expected.

Saeris commented:
> Left a couple of nits, but tested on staging and adding/changing/deleting secrets works as expected.

did you check the related PRs for adding/deleting? because this one only handles edits

Saeris committed:
update testing setup

Saeris committed:
Merge branch 'master' into chore/update-studio-testing-setup

joshenlim commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/SecretRow.tsx:
```diff
@@ -95,31 +104,24 @@ const SecretRow = ({ secret, onSelectEdit, onSelectRemove }: SecretRowProps) =>
-                <DropdownMenuItem
-                  className="space-x-2"
+
+        <Tooltip>
```
same thing RE consistency - we aren't currently following a pattern of disabling the dropdown menu button here

i'd stick to keeping the dropdown menu clickable, but disabling the menu items with DropdownMenuItemTooltip instead ([example](https://github.com/supabase/supabase/blob/master/apps/studio/components/interfaces/Auth/Policies/PolicyTableRow/PolicyRow.tsx#L152))

this is probably worth relooking as a pattern across the dashboard, but i'd say we do that separately

joshenlim commented on apps/studio/components/interfaces/Integrations/Vault/Secrets/EditSecretModal.tsx:
```diff
@@ -1,158 +1,214 @@
-      onCancel={onClose}
-      header={<h5 className="text-sm text-foreground">Edit secret</h5>}
+    <Dialog
+      open
```
one thing i noticed as well if we do this way, is that we won't have the animation of the modal when its closing (since the component is being unmounted immediately)

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
implement requested changes

Saeris commented:
@joshenlim @saltcod I've implemented all of the requested changes I think

simplified the test to remove the extra boilerplate, as after writing a handful more like it this week it doesn't seem to be necessary now that I'm more familiar with the dialog triggering patterns we're using

Saeris committed:
cleanup unnecessary conditional logic for undefined `secret` prop

Saeris committed:
remove unused import

Saeris committed:
remove unnecessary code

Saeris commented:
merged up and implemented the requested changes into the downstream PRs

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
cleanup, replace lodash import with JS

Saeris commented:
synced with master and did some minor cleanup. pretty sure all feedback has been addressed and changes implemented. let's merge?

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Merge branch 'chore/update-studio-testing-setup' into chore/refactor-edit-vault-secret-modal

Saeris committed:
add required router mock

Saeris commented on apps/studio/tests/lib/route-mock.ts:
```diff
@@ -1,11 +1,26 @@
-    '/projects/[ref]',
-  ])
-)
+const pipe =
+  <T>(...fns: Array<(arg: T) => T>) =>
+  (value: T) =>
+    fns.reduce((acc, fn) => fn(acc), value)
+// normalize file paths to posix format (windows FS support)
+const toPosix = (str: string) => str.replaceAll(sep, posix.sep)
+const pagesRoot = toPosix(join(process.cwd(), `./pages`))
+// remove the base filepath
+const removePrefix = (str: string) => str.replace(pagesRoot, ``)
+// remove the file extension or index segment
+const removeExt = (str: string) => str.replace(/.tsx|\/index.tsx/, ``)
+
+// Glob the `pages/` directory for all dynamic route segments
+const paths = [
+  ...(await Array.fromAsync(glob(`${pagesRoot}/**/*\].tsx`))),
+  ...(await Array.fromAsync(glob(`${pagesRoot}/**/**\]/**/index.tsx`))),
+].map(pipe(normalize, toPosix, removePrefix, removeExt))
+
+routerMock.useParser(createDynamicRouteParser(paths))
```
I thought this had already been merged in with another PR, but it must still be under review

without providing a list of our dynamic routes to parse, we can't extract a value like `projectRef` when `useParams` is called

since we don't throw any errors on this, many of our hooks will just accept an `undefined` value for proectRef and get stuck in a perpetual loading state, causing a test to fail due to a UI not updating or a by timing out

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

saltcod committed:
Merge branch 'master' into chore/refactor-edit-vault-secret-modal

Saeris committed:
fix broken lockfile

saltcod reviewed (approved):
Looks good on staging!
